src/ProductImage/Repositories/ProductImageRepository.php

```diff
@@ -31,4 +42,32 @@ public function modelClass(): string
     {
         return ProductImage::class;
     }
+
+    public function update(int $id, array $attributes): bool
+    {
+        $model = $this->model()->find($id);
+        $oldFile = $model->src;
+
+        if ($model->update($attributes)) {
+            $this->filesystem->disk('public')->delete($oldFile);
+
+            return true;
+        }
+
+        return false;
+    }
```

Comment:
Is it possible that the old *and* new `src` have the same value? Even if so, I suppose that's okay as long as the image is moved into place only after the model is updated (otherwise, we risk deleting the new image of the same name).

Also, I feel like a try/catch is more appropriate here, because currently, the method could still return `true` even if the file deletion fails, which could cause problems further up the stack.

Reply:
Yea, i was back and forth on this, was thinking about how to use a database transaction, problem is that when working with the filesystem there really isn't an "undo" method ... so its like one or the other will have to risk failing.

We could simply return on the file delete for example:

```php
public function update(int $id, array $attributes): bool
    {
        $model = $this->model()->find($id);
        $oldFile = $model->src;
        if ($model->update($attributes)) {
            return $this->filesystem->disk('public')->delete($oldFile);
        }
        return false;
    }
```

because that line does not throw an exception 
Ref: <https://github.com/laravel/framework/blob/6.x/src/Illuminate/Filesystem/Filesystem.php#L210> 
so a try/catch wouldn't do us much good here.

Oh i do have an idea to make transactions work tho, let me commit something and see what your thoughts are.

Reply:
Yes, at a minimum, we need a transaction here so that if the model update/delete succeeds, but the file deletion fails, we can rollback the model record.

In that vein, it's important to note that the database is the authoritative source here; in other words, the database entry is the "master record" that informs the application of that file's existence (rather than the file's presence on the filesystem). Therefore, the file should never be modified or deleted until the database record is modified or deleted, which is how you've structured the code already, so that aspect is solid as is.

Good point about `delete()` not throwing an exception on failure (I don't like Laravel's approach here, particularly with the `@` operator).

In any case, adding the transaction is probably adequate.

Reply:
The subsequent commit looks much better! This is resolved as far as I'm concerned.